### ISSUE TITLE
Group of Chains Exception/Error Propagation

### ISSUE THREAD

**Issue by br1anjtuck** [celery/celery#3810]
## Checklist
-  I'm running Celery 4.0.2

## Steps to reproduce
tasks.py
```
@shared_task
def add(x,y):
    return x+y

@shared_task
def raising_exception():
    raise RuntimeError("BLAH")
```
test.py
```
from celery import chord, group
from my_app.tasks import xsum, raising_exception

tasks = group(add.si(1,2), raising_exception.si() | add.si(1,1) )
result = tasks.apply_async()

try:
    result.get()
except Exception as ex:
    raise ex
```
Run the following:
celery worker -A grizzly_project -c 8 &
python manage.py runserver

then in another terminal, run:
python test.py
## Expected behavior
When the "raising_exception" task fails, result.get() should propagate one task back in the chain from the last task (for the second part of the group) to receive the exception: " RuntimeError: BLAH" thus allowing me to catch the Exception similar to below:
```
Traceback (most recent call last):
  File "test.py", line 15, in <module>
    raise ex
celery.backends.base.RuntimeError: BLAH
```
This is the expected behaviour, according to the back propagation of errors in a chain that was implemented in the closing of Issue [1014](https://github.com/celery/celery/issues/1014).

## Actual behavior
Rather, when the result.get() statement executes, the program hangs. 

When I change the group statement from the above code to this such that the failing task is the last to execute:
```
tasks = group(add.si(1,2), add.si(1,1) | raising_exception.si() )
```
...the result.get() doesn't hang but rather the Exception is caught, as expected 

Also, if i simply make it just a chain (rather than a group of chains) with the failing task first:
```
tasks = raising_exception.si() | add.si(1,1)
```
...result.get() doesn't hang but accurately catches the Exception (granted a wait a second or two before executing result.get())

**Comment by jathanism** [comment]
I would just like to throw my hat into the ring here because I'm encountering the same issue. So long as there are no errors, everything works as expected, but with errors, indefinite hang.

**Comment by snstanton** [comment]
Does anyone have a workaround for this?  Without this, there is no way to do exception handling in canvas and we have to catch all exceptions and propagate them in the results instead.

**Comment by OskarPersson** [comment]
I'm having the same problem given the test provided above using the following setup (master):

```
software -> celery:4.2.0 (windowlicker) kombu:4.2.2-post1 py:3.6.7
            billiard:3.5.0.5 py-amqp:2.4.0
platform -> system:Linux arch:64bit, ELF
            kernel version:4.9.125-linuxkit imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:pyamqp results:redis://redis/

broker_url: 'amqp://guest:********@rabbit:5672//'
result_backend: 'redis://redis/'
```

**Comment by SharpEdgeMarshall** [comment]
Same here with celery 4.3.0, I would like to add that this is really a critical issue.
I don't understand how is possible given that it should have already been fixed here https://github.com/celery/celery/commit/e043b82827874ed1c904e47b8dc964729b40442a

**Comment by auvipy** [comment]
could you try that patch locally?

**Comment by SharpEdgeMarshall** [comment]
What patch? that commit is already on master

**Comment by rusnyder** [comment]
This was not fixed by the aforementioned PR (https://github.com/celery/celery/pull/5587).

More specifically, the issue that still exists is that if any task in a chain that feeds into a chord _other than_ the very last task of that chain, then the error will not propagate to either the `AsyncResult.get()` method of the chord (which will hang) nor to the `AsyncResult.ready()` method of the chord, which will return `False` indefinitely.

I believe this is because there exists some parent-inspecting the result fetching code, but it doesn't account for chains.  I'm going to take a stab at what I believe to be as a passable solution to this and put out a PR.

**Comment by thedrow** [comment]
@rusnyder Is this test testing the wrong thing?
https://github.com/celery/celery/pull/5587/files#diff-2ae8afebeb9ba0fd1534a70264a2ac68R59

Can you provide the other test cases? Preferably in a form of a PR to our integration suite.

**Comment by SharpEdgeMarshall** [comment]
@rusnyder could you tell us what version of `vine` are you using?

**Comment by rusnyder** [comment]
Apologies for the delay, but I'll be getting back around to this in the next few days, sharing more info, and adding a test to the integration suite.

**Comment by ngrj93** [comment]
> This was not fixed by the aforementioned PR (#5587).
> 
> More specifically, the issue that still exists is that if any task in a chain that feeds into a chord _other than_ the very last task of that chain, then the error will not propagate to either the `AsyncResult.get()` method of the chord (which will hang) nor to the `AsyncResult.ready()` method of the chord, which will return `False` indefinitely.
> 
> I believe this is because there exists some parent-inspecting the result fetching code, but it doesn't account for chains. I'm going to take a stab at what I believe to be as a passable solution to this and put out a PR.

Was this problem solved because I am facing this exact same issue

**Comment by thedrow** [comment]
Not to the best of my knowledge.
Please provide a test case.
